### ISSUE TITLE
Elementwise assignments of intermediate tensors (based on parameter-dependent values)

### ISSUE THREAD

**Issue by andreaskoepf** [twitter/torch-autograd#77]
I would like to calculate my scalar result based on intermediate tensors that are filled with parameter-dependent values. An example would be to generate a rotation matrix from axis and angle parameters while being interested in the gradient w.r.t. axis and angle...

Unfortunately even the simplest things I try fail as soon as my input parameters are 'passed' through intermediate tensors. It seems as if autograd treats my intermediate tensors simply as constants which do not depend on the parameters but actually all of their elements are assigned from computations that are functions of the parameters...

Trivial example: 

```
function f(W)
  local x = torch.zeros(1)
  x[1] = torch.sin(W[1])
  return torch.sum(torch.abs(x))
end
df = grad(f)
dparams, loss = df(torch.ones(1))
```

Fails with `A node type was not returned. This is either because a gradient was not defined, or the input is independent of the output`...

Is the related to https://github.com/twitter/torch-autograd/issues/3 and new-index not being implemented yet or is this deliberately not supported?

**Comment by alexbw** [comment]
I agree that this is currently a frustrating aspect of autograd (fwiw, all versions of autograd in other languages suffer from this, too).  

The heart of this I believe is that `__newindex` and `__index` are not implemented, and in order to do so, we'd need to think hard about how to keep track of every element in a matrix that was overwritten, in case they were used for gradient calculations. I think `__index` is easy (we've already got support for `narrow` and `select`), we just haven't done it. For `__newindex`, naively, we could save the whole matrix, but we should probably be smart and not keep track of overwritten data that was never used in another calculation (as in your example). Honestly, I've been waiting for the mythical sparse tensor type to be finished (cc @soumith @clementfarabet), which would vastly simplify the implementation of this feature (just store what was overwritten in a sparse matrix, as opposed to the whole matrix, or in some hacked version of a sparse matrix)

Since this issue keeps coming up (both externally and internally), I think I'll start working on this, @luketwitter. 

Btw, what we do to get around this, we store pieces of tensors in tables, and use `torch.cat` to glue them together when needed. I agree assigning directly would be a great deal more convenient.

**Comment by andreaskoepf** [comment]
@alexbw I will try to understand a little bit more about the tracing mechanism which builds the graph for the backward pass. I could imagine that a challenging part of newindex is that it effectively does the tensor-change as a side-effect, in a pure functional world it would have to return the updated tensor... I am not sure how autograd deals with side-effects.

I am personally looking for two feature:
1. Automatic differentiation of operations on scalar values. e.g. the torch math functions also work for number-values but autograd expects always tensor arguments (e.g. the gradient output is passed as a param and this does not work for scalar values. Currently it would be necessary to use single element tensors to emulate scalar values (which of couse works but is a litter cumbersome..).
2. Tracking sources of tensor element assignments.

I think autograd is very interesting and I would love to help but I am busy until mid of February...

(sorry for typos - I am writing on the go on my mobile phone)

**Comment by alexbw** [comment]
1. You have it right. We need to stay purely functional, and assignment has
side effects. We're thinking of writing a utility function like util.set(x,
1, newVal) which would return a new tensor with the value set. Still
planning the feature.
2. This is an issue with torch itself. I think it's worth discussing the
issue over on their repository. When they support numbers, we will do so
"for free".

**Comment by alexbw** [comment]
Hey, I just added a utility function `util.cat`, which can concatenate numbers. If you want to build an affine transform matrix, try something like

```lua
local autograd = require 'autograd'
local affineMat = torch.reshape(autograd.util.cat({1,2,3,4,5,6}), 2, 3)
```

Concatenating numbers produces the default tensor type, so if you want float or cuda, just cast it afterwards.

Let me know if that helps you.

**Comment by alexbw** [comment]
Also, we now support taking derivatives w.r.t. scalar inputs. See the `CatNumber` test.

**Comment by alexbw** [comment]
I'm going to close this issue for now, since `torch.cat` should get you all the way in your particular use-case. 

If you find yourself really in need of general-purpose assignment like `a[x] = b`, please re-open this issue with a specific test-case that cannot be efficiently written using `torch.cat`. We will put off working on direct assignment as long as possible, because any implementation will be extremely inefficient without a great deal of effort (almost certainly requires writing a sparse tensor type), and we have no internal use for it at Twitter right now, lowering its priority.

**Comment by OptimusLime** [comment]
Is there a more explicit example of using Cuda tensors with autograd.util.cat? I don't think that actually works, unless I'm missing something. 

Here is a simple augmented version of the CatNumber test that appears to have an issue when Cuda Tensors are used: 

```lua
CatNumberCuda = function() 
      local function f(params, input)
         local tbl = {}
         tbl[#tbl+1] = params.a
         tbl[#tbl+1] = params.b
         tbl[#tbl+1] = params.c
         local a = autograd.util.cat(tbl)
         return -a*input
      end
      local df = autograd(f)
      local params = {a=1,b=2,c=3}  

      local input
      local useCuda = true
      if useCuda then 
         input = torch.CudaTensor(3):uniform()
         params = autograd.util.cast(params, 'cuda')
      else
         input = torch.Tensor(3):uniform()
      end

      local grads, loss = df(params, input)
      -- It just needs to run, gradcheck doesn't support numbers right now
end,
```

**Comment by alexbw** [comment]
If you take out autograd completely, the test doesn't work. When you are concatenating a table of numbers, autograd will default to the default Tensor type, which you will have to cast to CUDA.

**Comment by alexbw** [comment]
Btw @andreaskoepf, explicit assignment is now supported via #85. I'll add some examples, soon.
